rclpy/src/rclpy/context.cpp

```diff
@@ -153,9 +154,12 @@ Context::shutdown()
       std::remove(g_contexts.begin(), g_contexts.end(), rcl_context_.get()),
       g_contexts.end());
```

Comment:
With this PR there's a difference between how `Context::shutdown()` and `rclpy::shutdown_contexts()` treat `g_contexts`. The first one treats the global variable as a collection of all valid contexts, while the second treats it as a collection of all created contexts. I'd recommend picking one or the other and documenting which in a comment on `g_contexts`.


If `g_contexts` is all contexts, I'd recommend moving this line to the destructor `Context::~Context()`. That has an advantage that there's help from the language to make sure the global variable doesn't get leaked, but means contexts may live in that list for longer.

If `g_contexts` is all valid contexts, this removal is fine but I'd recommend going with your other solution of clearing `g_contexts` in `rclpy::shutdown_contexts()`.